File: Android/Paging-CRUD/app/src/main/java/com/hhp227/paging_crud/data/PostRemoteMediator.kt
package com.hhp227.paging_crud.data

import androidx.paging.ExperimentalPagingApi
import androidx.paging.LoadType
import androidx.paging.PagingState
import androidx.paging.RemoteMediator
import com.hhp227.paging_crud.api.PostService
import com.hhp227.paging_crud.model.ListItem
import kotlinx.coroutines.delay
import retrofit2.HttpException
import java.io.IOException

@OptIn(ExperimentalPagingApi::class)
class PostRemoteMediator(
    private val postService: PostService,
    private val postDao: PostDao,
    private val groupId: Int
) : RemoteMediator<Int, ListItem.Post>() {
    override suspend fun load(loadType: LoadType, state: PagingState<Int, ListItem.Post>): MediatorResult {
        return try {
            val offset = when (loadType) {
                LoadType.REFRESH -> 0
                LoadType.PREPEND -> return MediatorResult.Success(endOfPaginationReached = true)
                LoadType.APPEND -> {
                    // 하단 도달시 로딩 인디케이터가 잠시 보이도록 의도적으로 지연
                    delay(1000)
                    postDao.getCount(groupId)
                }
            }
            val loadSize = if (loadType == LoadType.REFRESH) state.config.initialLoadSize else state.config.pageSize
            val response = postService.getPostList(groupId, offset, loadSize)

            if (!response.error) {
                val data = response.data ?: emptyList()

                if (loadType == LoadType.REFRESH) {
                    postDao.replaceAll(groupId, data)
                } else {
                    postDao.insertAll(groupId, data)
                }
                MediatorResult.Success(endOfPaginationReached = data.isEmpty())
            } else {
                MediatorResult.Error(
                    Throwable(response.message)
                )
            }
        } catch (e: IOException) {
            MediatorResult.Error(e)
        } catch (e: HttpException) {
            MediatorResult.Error(e)
        }
    }
}

File: iOS/Paging-CRUD/Paging-CRUD/Data/PostRemoteMediator.swift
//
//  PostRemoteMediator.swift
//  Paging-CRUD
//

import Foundation
import Paging

final class PostRemoteMediator: RemoteMediator<Int, ListItem.Post> {
    private let postService: PostService

    private let postDao: PostDao

    private let groupId: Int

    init(postService: PostService, postDao: PostDao, groupId: Int) {
        self.postService = postService
        self.postDao = postDao
        self.groupId = groupId
        super.init()
    }

    override func load(loadType: LoadType, state: PagingState<Int, ListItem.Post>) async -> MediatorResult {
        do {
            let offset: Int

            switch loadType {
            case .refresh:
                offset = 0
            case .prepend:
                return .success(endOfPaginationReached: true)
            case .append:
                // 하단 도달시 로딩 인디케이터가 잠시 보이도록 의도적으로 지연
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                offset = postDao.getCount(groupId)
            }
            let loadSize = loadType == .refresh ? state.config.initialLoadSize : state.config.pageSize
            let response = try await postService.getPostList(groupId: groupId, offset: offset, loadSize: loadSize)

            if !response.error {
                let data = response.data ?? []

                if loadType == .refresh {
                    postDao.replaceAll(groupId, data)
                } else {
                    postDao.insertAll(groupId, data)
                }
                return .success(endOfPaginationReached: data.isEmpty)
            } else {
                return .error(
                    NSError(
                        domain: "PostRemoteMediator",
                        code: 0,
                        userInfo: [NSLocalizedDescriptionKey: response.message ?? "An unexpected error occured"]
                    )
                )
            }
        } catch {
            return .error(error)
        }
    }
}
